core/src/memory/instrumenting_memory_resource.cpp

```diff
@@ -0,0 +1,173 @@
+
+    std::size_t time =
+        std::chrono::duration_cast<std::chrono::nanoseconds>(t2 - t1).count();
+
+    /*
+     * Add a new allocation event with the size, alignment, pointer, and time
+     * of what has just happened.
+     */
+    m_events.emplace_back(memory_event::type::ALLOCATION, size, align, ptr,
+                          time);
+
+    /*
+     * Now, we can run the post-allocation hooks. For failed allocations, the
+     * pointer will be null.
+     */
+    for (const std::function<void(std::size_t, std::size_t, void *)> &f :
+         m_post_allocate_hooks) {
+        f(size, align, ptr);
+    }
+
+    /*
+     * Now we check whether our allocation failed. If that is the case, we just
+     * throw another bad allocation exception.
+     */
+    if (ptr == nullptr) {
+        throw std::bad_alloc();
+    } else {
+        /*
+         * If the allocation was a success, add the size to the total
+         * outstanding memory pool.
+         */
+        m_outstanding += size;
+    }
+
+    return ptr;
+}
+
+void instrumenting_memory_resource::do_deallocate(void *ptr, std::size_t size,
+                                                  std::size_t align) {
+    /*
+     * First, we will run all of our pre-deallocation hooks.
+     */
+    for (const std::function<void(void *, std::size_t, std::size_t)> &f :
+         m_pre_deallocate_hooks) {
+        f(ptr, size, align);
+    }
+
+    /*
+     * Now, remove the requested amount of memory from the outstanding pool. We
+     * don't really have a notion of deallocation failing, so we always assume
+     * that it succeeds. Calling deallocate on memory that was not succesfully
+     * allocated is UB anyway.
+     */
+    m_outstanding -= size;
+
+    /*
+     * As with allocation, we calculate the time taken to process this
+     * deallocation.
+     */
+    std::chrono::high_resolution_clock::time_point t1 =
+        std::chrono::high_resolution_clock::now();
+
+    /*
+     * The deallocation, like allocation, is a forwarding method.
+     */
+    m_upstream.deallocate(ptr, size, align);
+
+    /*
+     * Compute the total elapsed time during the deallocation request.
+     */
+    std::chrono::high_resolution_clock::time_point t2 =
+        std::chrono::high_resolution_clock::now();
+
+    std::size_t time =
+        std::chrono::duration_cast<std::chrono::nanoseconds>(t2 - t1).count();
+
+    /*
+     * Register a deallocation event.
+     */
+    m_events.emplace_back(memory_event::type::DEALLOCATION, size, align, ptr,
+                          time);
+}
+
+bool instrumenting_memory_resource::do_is_equal(
+    const memory_resource &other) const noexcept {
+    /*
+     * An instrumenting memory resource is only ever equal to itself.
+     * Technically this is not true, because if the upstreams are the same then
+     * they are equal. However, this cannot be allowed to happen because it
+     * would break our administration
+     */
+    return this == &other;
+}
```

Comment:
You could leave this one out. `vecmem::details::memory_resource_base` implement this function in the same way.

Reply:
Fixed. :smile: